src/wallet/wallet.cpp

```diff
@@ -3936,9 +3936,11 @@ CKeyPool::CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn)
 
 int CWalletTx::GetDepthInMainChain(interfaces::Chain::Lock& locked_chain) const
 {
+    assert(pwallet != nullptr);
+    AssertLockHeld(pwallet->cs_wallet);
     if (isUnconfirmed() || isAbandoned()) return 0;
 
-    return locked_chain.getBlockDepth(m_confirm.hashBlock) * (isConflicted() ? -1 : 1);
```

Comment:
In commit "Use CWallet::m_last_block_processed_height in GetDepthInMainChain" ([0ff0387](https://github.com/bitcoin/bitcoin/commit/0ff03871add000f8b4d8f82aeb168eed2fc9dc5f))

Dropping the `getBlockDepth` call here introduces a bug because `getBlockDepth` had a safety check to return 0 if there was a reorg and hashBlock is no longer in the active chain, while the new code will just return a garbage depth. I think this is also the reason adding height asserts didn't work in some tests: https://github.com/bitcoin/bitcoin/pull/15931#discussion_r371512762. A few fixes are possible and mentioned [*bug-reorg-depth*](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking#bug-reorg-depth). A fix shouldn't be hard, but a bigger challenge is adding better test coverage to prevent more bugs like this.